File: apps/fleet-monitor/fleet_monitor/tasks.py
import logging
from collections.abc import Iterable

log = logging.getLogger("fleet.tasks")


def log_raised[T](label: str, outcomes: Iterable[T | BaseException]) -> tuple[T, ...]:
    """Name any job that raised and return only the ones that produced a value.

    gather(return_exceptions=True) is what keeps one wedged host from taking a
    whole round down, but a swallowed exception is a silent hole, so every one
    of them is logged here before being dropped.

    Only `Exception` is swallowed. A BaseException that is not one is the
    process being torn down - CancelledError above all - and absorbing that
    makes graceful shutdown impossible, so it is re-raised instead.

    Shared by the vitals loop and the play-history loop, which run side by side
    in one process and must treat a torn-down task the same way.
    """
    collected = tuple(outcomes)
    fatal = next(
        (
            item
            for item in collected
            if isinstance(item, BaseException) and not isinstance(item, Exception)
        ),
        None,
    )
    if fatal is not None:
        raise fatal
    raised = tuple(item for item in collected if isinstance(item, Exception))
    if raised:
        log.warning(
            "%s: %d job(s) raised: %s", label, len(raised), "; ".join(map(repr, raised))
        )
    return tuple(item for item in collected if not isinstance(item, Exception))

File: apps/fleet-monitor-nest/src/tasks.ts
import { Logger } from '@nestjs/common'

const log = new Logger('fleet.tasks')

// An abort is the process being torn down, not a job failing, and swallowing
// it would make a graceful shutdown impossible.
const isAbort = (reason: unknown): boolean =>
  reason instanceof Error && reason.name === 'AbortError'

/**
 * Name any job that rejected and return only the values the others produced.
 *
 * Promise.allSettled is what keeps one wedged host from taking a whole round
 * down, but a swallowed rejection is a silent hole, so every one of them is
 * logged here before being dropped. An abort is rethrown instead.
 *
 * Shared by the vitals loop and the play-history loop, which run side by side
 * in one process and must treat a torn-down task the same way.
 */
export const logRaised = <T>({
  label,
  outcomes,
}: {
  label: string
  outcomes: readonly PromiseSettledResult<T>[]
}): T[] => {
  const reasons = outcomes.flatMap((outcome) =>
    outcome.status === 'rejected' ? [outcome.reason] : [],
  )
  const fatal = reasons.find(isAbort)
  if (fatal !== undefined) {
    throw fatal
  }
  if (reasons.length > 0) {
    log.warn(`${label}: ${reasons.length} job(s) raised: ${reasons.map(String).join('; ')}`)
  }
  return outcomes.flatMap((outcome) => (outcome.status === 'fulfilled' ? [outcome.value] : []))
}
